### PR TITLE
Update IPAllow infrastructure.

### PR DESCRIPTION
**Pull request by SolidWallOfCode**
This was motivated by a number of crashes in Oath production. To some extent it was fallout from the problem in #4122 but there was still a problem in IPAllow itself. The root was that when a `ProxyClientSession` gets the ACL record, that is a raw pointer in to an element in an array in the configuration instance. If the configuration is reloaded, that pointer goes stale. Without the multiple reload issue this happens extremely rare because the ACL record is used only briefly.

Fixing this requires keeping a reference count to the config along with the ACL. To do that, I added a new wrapper class on top of the ACL record pointer which also carries a config pointer so the reference can be cleared when the ACL owner gets shut down. I also fixed a crash that occurred when the debug tag was set and an ACL with a non-standard method was denied.

A side effect of this was that the only method in `SessionAccept.cc` was removed so I removed the entire file.

I ended up redoing the parsing to get better control of memory usage. The interface was changed to use `string_view` so that checking didn't require creating a `std::string` and its attendant internal allocation. The easiest way to make that work was to keep the original file contents around after parsing and have the `string_view` instances point in to that. Additionally I moved the other method data into a `std::vector` instead of a `std::map` because the number of custom methods is in practice never more than 2 or 3 and a vector is not only better memory but faster in such cases. The parsing error messages were also improved to provide more detail on errors.

This requires #4115.

### CODE REVIEW

**Comment by SolidWallOfCode** [comment]
[approve ci freebsd]

**Comment by SolidWallOfCode** [comment]
[approve ci autest]

File: proxy/IPAllow.cc [hunk 1]
```diff
@@ -81,47 +88,70 @@ IpAllow::reconfigure()
...
+    }
+  } else {
+    self->_dst_map.contains(ip, &raw);
+  }
+  if (raw == nullptr) {
+    self->release();
+    self = nullptr;
+  }
+  return {static_cast<Record *>(raw), self};
 }
```
Comment (shinrich) [inline proxy/IPAllow.cc]: So this is a constructor?

File: proxy/IPAllow.cc [hunk 2]
```diff
@@ -81,47 +88,70 @@ IpAllow::reconfigure()
...
 
 void
-IpAllow::release(IpAllow *lookup)
+IpAllow::release()
+{
+  configProcessor.release(configid, this);
+}
+
+auto
+IpAllow::match(sockaddr const *ip, match_key_t key) -> ACL
```
Comment (shinrich) [inline proxy/IPAllow.cc]: To me anyway, replacing auto with IpAllow::ACL would be more clear than the "-> ACL".  The new syntax isn't really more concise.

File: proxy/IPAllow.cc (hunk 1 above)
Comment (SolidWallOfCode) [reply proxy/IPAllow.cc]: Yes. The compiler knows the return type and so automatically coerces the initializer list in to the returned object type.

File: proxy/IPAllow.cc (hunk 2 above)
Comment (SolidWallOfCode) [reply proxy/IPAllow.cc]: I've been working with this notation, and the rule I was using was to declare the return type normally if it's an outer scope type, and via `auto` if it's a dependent scope type. It's nice in situations where the dependent type is deeply nested, like
```
GenericClass::NestedContainer::iterator::value_type
GenericClass::NestedContainer::iterator::operator * () { ... }
// vs.
auto
GenericClass::NestedContainer::iterator::operator * () -> value_type { ... }
```
I prefer a simple rule to discussions about exactly how many characters are enough to sufficiently concise.

**Pull request merged**